File: Uncategorized/hkccc15j4.cpp
#include <bits/stdc++.h>
#define f first
#define s second
#define sc(x) do{while((x=getchar())<'0'); for(x-='0'; '0'<=(_=getchar()); x=(x<<3)+(x<<1)+_-'0');}while(0)
char _;
using namespace std;
typedef pair<int,int> pii;
const int MM = 1e5+2;
pii a[MM];
int L,N;
int main(){
    sc(L); sc(N);
    for(int i = 0; i < N; i++){
        sc(a[i].f); sc(a[i].s);
    }
    sort(a, a+N);
    a[N] = {L, L};
    int pre = 0, ans = 0;
    for(int i = 0; i <= N; i++){
        if(pre < a[i].f)
            ans = max(ans, a[i].f - pre);
        pre = max(pre, a[i].s);
    }
    printf("%d\n", ans);
    return 0;
}

File: Uncategorized/hkccc15j4.py
import sys
input = sys.stdin.readline
l = int(input())
n = int(input())
arr = list()
for i in range(n):
	a, b = map(int, input().split())
	arr.append([a, b])

arr.append([l, l])
arr.sort()

ans = 0
r = 0
for i in arr:
	if i[0] <= r:
		r = max(r, i[1])
	else:
		ans = max(ans, i[0]-r)
		r = i[1]

print(ans)